src/Database/PostgreSQL/Simple/Interval.hs

```diff
@@ -0,0 +1,241 @@
+{-# LANGUAGE NumDecimals #-}
+
+module Database.PostgreSQL.Simple.Interval where
+
+import qualified Control.Applicative as Applicative
+import qualified Control.Monad as Monad
+import qualified Data.Attoparsec.ByteString.Char8 as A
+import qualified Data.Bits as Bits
+import qualified Data.ByteString as ByteString
+import qualified Data.ByteString.Builder as Builder
+import qualified Data.Function as Function
+import qualified Data.Int as Int
+import qualified Data.Scientific as Scientific
+import qualified Database.PostgreSQL.Simple.FromField as Postgres
+import qualified Database.PostgreSQL.Simple.ToField as Postgres
+import qualified Database.PostgreSQL.Simple.TypeInfo.Static as Postgres
+
+data Interval = MkInterval
+  { months :: !Int.Int32,
+    days :: !Int.Int32,
+    microseconds :: !Int.Int64
+  }
+  deriving (Eq, Show)
+
+instance Postgres.FromField Interval where
+  fromField = Postgres.attoFieldParser (== Postgres.intervalOid) parse
+
+instance Postgres.ToField Interval where
+  toField = Postgres.Plain . render
+
+zero :: Interval
+zero = MkInterval 0 0 0
+
+fromMicroseconds :: Int.Int64 -> Interval
+fromMicroseconds x = zero {microseconds = x}
+
+fromMilliseconds :: Int.Int64 -> Maybe Interval
+fromMilliseconds =
+  fmap fromMicroseconds
+    . Bits.toIntegralSized
+    . (*) 1e3
+    . toInteger
+
+fromSeconds :: Int.Int64 -> Maybe Interval
+fromSeconds =
+  fmap fromMicroseconds
+    . Bits.toIntegralSized
+    . (*) 1e6
+    . toInteger
+
+fromMinutes :: Int.Int64 -> Maybe Interval
+fromMinutes =
+  fmap fromMicroseconds
+    . Bits.toIntegralSized
+    . (*) 60e6
+    . toInteger
+
+fromHours :: Int.Int64 -> Maybe Interval
+fromHours =
+  fmap fromMicroseconds
+    . Bits.toIntegralSized
+    . (*) 3600e6
+    . toInteger
+
+fromDays :: Int.Int32 -> Interval
+fromDays x = zero {days = x}
+
+fromWeeks :: Int.Int32 -> Maybe Interval
+fromWeeks =
+  fmap fromDays
+    . Bits.toIntegralSized
+    . (*) 7
+    . toInteger
+
+fromMonths :: Int.Int32 -> Interval
+fromMonths x = zero {months = x}
+
+fromYears :: Int.Int32 -> Maybe Interval
+fromYears =
+  fmap fromMonths
+    . Bits.toIntegralSized
+    . (*) 12
+    . toInteger
+
+add :: Interval -> Interval -> Maybe Interval
+add x y =
+  let safeAdd :: (Bits.Bits a, Integral a) => a -> a -> Maybe a
+      safeAdd n = Bits.toIntegralSized . Function.on (+) toInteger n
+   in MkInterval
+        <$> Function.on safeAdd months x y
+        <*> Function.on safeAdd days x y
+        <*> Function.on safeAdd microseconds x y
+
+render :: Interval -> Builder.Builder
+render x =
+  let signed :: (Num a, Ord a) => (a -> Builder.Builder) -> a -> Builder.Builder
+      signed f n = (if n > 0 then "+" else "") <> f n
+   in "interval '"
+        <> signed Builder.int32Dec (months x)
+        <> " months "
+        <> signed Builder.int32Dec (days x)
+        <> " days "
+        <> signed Builder.int64Dec (microseconds x)
+        <> " microseconds'"
+
+parse :: A.Parser Interval
```

Comment:
This is not a general purpose parser for Postgres intervals. It only does enough to parse intervals that Postgres emits. 

It is too restrictive. For example, none of the parsers handle weeks because Postgres never renders intervals using weeks. 

It is also too permissive. For example, it allows components to be repeated when that should sometimes be an error.